File: frontend/android/app/src/main/java/com/quietcue/app/domain/AcousticFingerprint.kt
package com.quietcue.app.domain

import java.util.UUID
import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.log10
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

data class CapturedFingerprint(
    val features: List<Float>,
    val rmsDbfs: Float,
)

object AcousticFingerprint {
    const val SAMPLE_RATE = 16_000
    const val FEATURE_COUNT = 8
    const val MIN_POSITIVE_SAMPLES = 3
    const val RECOMMENDED_POSITIVE_SAMPLES = 6
    const val MAX_POSITIVE_SAMPLES = 30
    val frequencies = intArrayOf(250, 375, 500, 750, 1_000, 1_500, 2_000, 3_000)

    fun fromPcm16(samples: ShortArray): CapturedFingerprint {
        require(samples.isNotEmpty()) { "Cannot fingerprint empty audio" }
        val rms = sqrt(samples.sumOf { sample -> sample.toDouble() * sample } / samples.size) / 32768.0
        val amplitudes = frequencies.map { frequency -> goertzelAmplitude(samples, frequency) }
        val norm = sqrt(amplitudes.sumOf { value -> value * value }).coerceAtLeast(1e-12)
        return CapturedFingerprint(
            features = amplitudes.map { (it / norm).toFloat() },
            rmsDbfs = (20.0 * log10(max(rms, 1e-9))).toFloat(),
        )
    }

    fun similarity(left: List<Float>, right: List<Float>): Float {
        if (left.size != FEATURE_COUNT || right.size != FEATURE_COUNT) return 0f
        val dot = left.indices.sumOf { index -> left[index].toDouble() * right[index] }
        val leftNorm = sqrt(left.sumOf { it.toDouble() * it })
        val rightNorm = sqrt(right.sumOf { it.toDouble() * it })
        return (dot / max(leftNorm * rightNorm, 1e-12)).toFloat().coerceIn(0f, 1f)
    }

    fun enroll(
        name: String,
        description: String,
        priority: AlertPriority,
        positives: List<CapturedFingerprint>,
        background: CapturedFingerprint,
        nowMs: Long = System.currentTimeMillis(),
        confusingSounds: List<CapturedFingerprint> = emptyList(),
    ): SoundDefinition {
        require(positives.size in MIN_POSITIVE_SAMPLES..MAX_POSITIVE_SAMPLES) {
            "Record between $MIN_POSITIVE_SAMPLES and $MAX_POSITIVE_SAMPLES examples"
        }
        require(positives.all { it.rmsDbfs >= -50f }) { "One recording is too quiet; record it again" }
        val averaged = List(FEATURE_COUNT) { index -> positives.map { it.features[index] }.average() }
        val norm = sqrt(averaged.sumOf { it * it }).coerceAtLeast(1e-12)
        val prototype = averaged.map { (it / norm).toFloat() }
        val negativeSamples = listOf(background) + confusingSounds
        val backgroundSimilarity = negativeSamples.maxOf { negative ->
            positives.maxOf { positive -> similarity(positive.features, negative.features) }
        }
        require(backgroundSimilarity < 0.92f) {
            "A background or confusing sound is too similar to an example; replace that recording"
        }
        val threshold = max(0.80f, backgroundSimilarity + 0.04f).coerceAtMost(0.95f)
        val pattern = when (priority) {
            AlertPriority.INFORMATIONAL -> HapticPattern.TWO_SHORT
            AlertPriority.ATTENTION -> HapticPattern.LONG_PULSE
            AlertPriority.EMERGENCY -> HapticPattern.URGENT_REPEAT
        }
        return SoundDefinition(
            id = "custom:${UUID.randomUUID()}",
            displayName = name.trim(),
            description = description.trim().ifBlank { "Sound enrolled with the phone microphone" },
            safetyCritical = priority == AlertPriority.EMERGENCY,
            defaultPriority = priority,
            defaultHapticPattern = pattern,
            defaultHapticStrength = if (priority == AlertPriority.EMERGENCY) {
                HapticStrength.STRONG
            } else {
                HapticStrength.STANDARD
            },
            defaultRequiresAcknowledgement = priority == AlertPriority.EMERGENCY,
            enrollment = SoundEnrollment(
                prototype = prototype,
                prototypes = positives.map(CapturedFingerprint::features),
                sampleRmsDbfs = positives.map(CapturedFingerprint::rmsDbfs),
                similarityThreshold = threshold,
                positiveSampleCount = positives.size,
                backgroundSimilarity = backgroundSimilarity,
                createdAtEpochMs = nowMs,
            ),
        )
    }

    private fun goertzelAmplitude(samples: ShortArray, frequency: Int): Double {
        val coefficient = 2.0 * cos(2.0 * PI * frequency / SAMPLE_RATE)
        var previous = 0.0
        var previousPrevious = 0.0
        samples.forEach { sample ->
            val current = sample + coefficient * previous - previousPrevious
            previousPrevious = previous
            previous = current
        }
        val power = previousPrevious * previousPrevious + previous * previous -
            coefficient * previous * previousPrevious
        return 2.0 * sqrt(max(0.0, power)) / samples.size
    }
}

File: frontend/ios/QuietCue/Domain/AcousticFingerprint.swift
import Foundation

struct CapturedFingerprint: Equatable, Hashable {
    var features: [Float]
    var rmsDbfs: Float
}

struct EnrollmentCapture: Equatable {
    var positives: [CapturedFingerprint]
    var background: CapturedFingerprint
    var speechRejectedMs: Int = 0
}

struct EnrollmentError: LocalizedError {
    var message: String
    var errorDescription: String? { message }
}

enum AcousticFingerprint {
    static let sampleRate = 16_000
    static let featureCount = 8
    static let minPositiveSamples = 3
    static let recommendedPositiveSamples = 6
    static let maxPositiveSamples = 30
    static let frequencies = [250, 375, 500, 750, 1_000, 1_500, 2_000, 3_000]
    private static let minRepeatSimilarity: Float = 0.86

    static func fromPcm16(_ samples: [Int16]) -> CapturedFingerprint {
        precondition(!samples.isEmpty, "Cannot fingerprint empty audio")
        let sumSquares = samples.reduce(0.0) { $0 + Double($1) * Double($1) }
        let rms = (sumSquares / Double(samples.count)).squareRoot() / 32768.0
        let amplitudes = frequencies.map { goertzelAmplitude(samples, frequency: $0) }
        let norm = max(amplitudes.reduce(0.0) { $0 + $1 * $1 }.squareRoot(), 1e-12)
        return CapturedFingerprint(
            features: amplitudes.map { Float($0 / norm) },
            rmsDbfs: Float(20.0 * log10(max(rms, 1e-9)))
        )
    }

    static func similarity(_ left: [Float], _ right: [Float]) -> Float {
        guard left.count == featureCount, right.count == featureCount else { return 0 }
        var dot = 0.0, leftNorm = 0.0, rightNorm = 0.0
        for index in 0..<featureCount {
            dot += Double(left[index]) * Double(right[index])
            leftNorm += Double(left[index]) * Double(left[index])
            rightNorm += Double(right[index]) * Double(right[index])
        }
        let value = dot / max(leftNorm.squareRoot() * rightNorm.squareRoot(), 1e-12)
        return Float(min(max(value, 0), 1))
    }

    static func matchConfidence(_ candidate: [Float], enrollment: SoundEnrollment) -> Float? {
        let references = enrollment.prototypes.isEmpty ? [enrollment.prototype] : enrollment.prototypes
        let similarities = references.map { similarity(candidate, $0) }.sorted(by: >)
        guard let best = similarities.first, best >= enrollment.similarityThreshold else { return nil }
        if enrollment.matcherVersion >= 4 && similarities.count >= minPositiveSamples {
            let tolerance: Float = 0.04
            if similarities[1] < enrollment.similarityThreshold - tolerance ||
                similarity(candidate, enrollment.prototype) < enrollment.similarityThreshold - tolerance {
                return nil
            }
        }
        return best
    }

    static func enroll(
        name: String,
        description: String,
        priority: AlertPriority,
        positives: [CapturedFingerprint],
        background: CapturedFingerprint,
        nowMs: Int64 = nowEpochMs(),
        confusingSounds: [CapturedFingerprint] = []
    ) throws -> SoundDefinition {
        guard (minPositiveSamples...maxPositiveSamples).contains(positives.count) else {
            throw EnrollmentError(message: "Record between \(minPositiveSamples) and \(maxPositiveSamples) examples")
        }
        guard positives.allSatisfy({ $0.rmsDbfs >= -50 }) else {
            throw EnrollmentError(message: "One recording is too quiet; record it again")
        }
        let averaged = (0..<featureCount).map { index in
            positives.map { Double($0.features[index]) }.reduce(0, +) / Double(positives.count)
        }
        let norm = max(averaged.reduce(0.0) { $0 + $1 * $1 }.squareRoot(), 1e-12)
        let prototype = averaged.map { Float($0 / norm) }
        let consistent = positives.filter { similarity($0.features, prototype) >= minRepeatSimilarity }.count
        guard consistent >= minPositiveSamples else {
            throw EnrollmentError(message: "The examples are too different; teach the same sound at least three times")
        }
        let negatives = [background] + confusingSounds
        let backgroundSimilarity = negatives.map { negative in
            positives.map { similarity($0.features, negative.features) }.max() ?? 0
        }.max() ?? 0
        guard backgroundSimilarity < 0.90 else {
            throw EnrollmentError(message: "A background or confusing sound is too similar to an example; replace that recording")
        }
        let threshold = min(max(0.93, backgroundSimilarity + 0.08), 0.98)
        let pattern: HapticPattern
        switch priority {
        case .informational: pattern = .twoShort
        case .attention: pattern = .longPulse
        case .emergency: pattern = .urgentRepeat
        }
        let trimmedDescription = description.trimmingCharacters(in: .whitespacesAndNewlines)
        return SoundDefinition(
            id: "custom:\(UUID().uuidString.lowercased())",
            displayName: name.trimmingCharacters(in: .whitespacesAndNewlines),
            descriptionText: trimmedDescription.isEmpty ? "Sound enrolled with the phone microphone" : trimmedDescription,
            safetyCritical: priority == .emergency,
            defaultPriority: priority,
            defaultHapticPattern: pattern,
            defaultHapticStrength: priority == .emergency ? .strong : .standard,
            defaultRequiresAcknowledgement: priority == .emergency,
            enrollment: SoundEnrollment(
                prototype: prototype,
                prototypes: positives.map(\.features),
                sampleRmsDbfs: positives.map(\.rmsDbfs),
                similarityThreshold: threshold,
                positiveSampleCount: positives.count,
                backgroundSimilarity: backgroundSimilarity,
                createdAtEpochMs: nowMs,
                matcherVersion: 4
            )
        )
    }

    private static func goertzelAmplitude(_ samples: [Int16], frequency: Int) -> Double {
        let coefficient = 2.0 * cos(2.0 * Double.pi * Double(frequency) / Double(sampleRate))
        var previous = 0.0
        var previousPrevious = 0.0
        for sample in samples {
            let current = Double(sample) + coefficient * previous - previousPrevious
            previousPrevious = previous
            previous = current
        }
        let power = previousPrevious * previousPrevious + previous * previous - coefficient * previous * previousPrevious
        return 2.0 * max(0.0, power).squareRoot() / Double(samples.count)
    }
}
